docs_src/design/adr/device-service/0012-DeviceService-Filters.md

```diff
@@ -0,0 +1,83 @@
+
+## Status
+
+### Proposed
+
+- design for Hanoi
+- implementation for Ireland
+
+## Context
+
+In EdgeX today, sensor/device data collected can be "filtered" by [application services](../../../microservices/application/ApplicationServices.md) before being exported or sent to some [north side](../../../general/Definitions.md#south-and-north-side) application or system. Built-in application service functions (available through the app services SDK) allow EdgeX event/reading objects to be filtered by device name or by value descriptor type.  That is, event/readings can be filtered by:
+
+- which device sent the event/reading (as determined by the device name in the event/readings)
+- the category (such as temperature or humidity) of data as determined by the value descriptor.
+
+*With value descriptor's deprecation, what will the app function SDK offer for filtering?*
+
+### Filter by device name and value descriptor nearer collection
+
+While application service filters help to limit the amount of data sent north, those events and readings are still collected, stored, processed by the EdgeX micro services, potentially resulting in wasted resources processing unneeded data.
+
+Therefore, there is a desire to help filter data at the device service level - thereby limiting the amount of sensor data that is captured, stored and processed as well as sent north.
+
+In the first stage of device service filtering, it is desired that the filter by device name and value descriptor be offered in all device services (via the SDK).  The filtering of event/readings should be accomplished just before the event/readings are sent to core data or (or put on the message bus in future implementations where DS message application services).  See Design below for more details.
+
+### Sharing filter functions
+
+If one were to explore the filtering functions in the app functions SDK [filter.go](https://github.com/edgexfoundry/app-functions-sdk-go/blob/master/pkg/transforms/filter.go) (both FilterByDeviceName and FilterByValueDescriptor), the filters operate on a simple event model object.  Ideally, since both app services and device services share the event model (from go-mod-core-contracts), it would be the desire to share filter functions to some degree between SDKs.
+
+### Future Filters
+
+There are additional characteristics by which device services may want to filter event/readings.  Since the device service originates the event/reading, there may be circumstances where by the event/reading is considered invalid or not valuable enough to even process through the rest of EdgeX.  So, in addition to filtering by particular devices or value descriptor, device services may in the future offer filters:
+
+- based on the reading value (numeric) of a reading outside a specified range (min/max) described in the device profile for a device resource.  Thus avoiding sending in outlier or jittery data readings that could negatively effect analytics.

```

Comment:
Yes! :-)

Reply:
And if the filter function module/library is our approach, these become available in AS too.  Why not?

Reply:
I checked our original device sdk requirements document and the only thing it says about `Minimum/Maximum` is:

_Minimum and Maximum are attributes in a device resource's PropertyValue which may be used to specify an allowable range for numeric actuation parameters. A device service should validate parameters against these before applying any transformations. An out-of-range request should result in a 400 error._

I see that some of the content from this original requirements document made it into the legacy folder in edgex-docs, but I don't see any details about `PropertyValue` (or `ValueDescriptor`) fields?

I think for future filters we could implement some builtin filters like this (i.e. that use `Min/Max` from a device resource's `PropertyValue`), but I think they should also be flexible enough that someone could say filter every reading with value < `Min`, or `value` <= `some specified value`.